File: hospital-core/src/main/java/com/example/hms/service/SmartPhraseService.java
package com.example.hms.service;

import com.example.hms.payload.dto.SmartPhraseRequestDTO;
import com.example.hms.payload.dto.SmartPhraseResponseDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * SmartPhrase / dot-phrase macro library used by the per-section EncounterNote
 * form. Resolution precedence at autocomplete time is USER > HOSPITAL > GLOBAL
 * — a user macro shadows a hospital one with the same trigger; a hospital
 * macro shadows a global one.
 *
 * <p>Mutations ({@link #create}, {@link #update}, {@link #delete}) read the
 * authenticated principal from the Spring Security context and enforce
 * scope-based authorization internally — a clinician with a known id cannot
 * edit another user's macros or create GLOBAL/HOSPITAL macros for hospitals
 * they do not staff.
 */
public interface SmartPhraseService {

    SmartPhraseResponseDTO create(SmartPhraseRequestDTO request);

    SmartPhraseResponseDTO update(UUID id, SmartPhraseRequestDTO request);

    void delete(UUID id);

    SmartPhraseResponseDTO get(UUID id);

    /** Page through GLOBAL phrases (admin library view). */
    Page<SmartPhraseResponseDTO> listGlobal(Pageable pageable);

    /**
     * Trigger-prefix autocomplete for the calling user at their active hospital.
     * Sorted USER > HOSPITAL > GLOBAL, then alphabetical by trigger.
     */
    List<SmartPhraseResponseDTO> autocomplete(String prefix, UUID hospitalId);

    /** Look up a single SmartPhrase by exact trigger for the caller. */
    Optional<SmartPhraseResponseDTO> findByTrigger(String trigger, UUID hospitalId);

    /** Increment usage_count + last_used_at — fire-and-forget from the FE. */
    void recordUsage(UUID id);
}

File: hospital-portal/src/app/services/smart-phrase.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

export type SmartPhraseScope = 'GLOBAL' | 'HOSPITAL' | 'USER';

export interface SmartPhrase {
  id: string;
  trigger: string;
  title: string;
  expansion: string;
  scope: SmartPhraseScope;
  hospitalId?: string;
  ownerUserId?: string;
  specialty?: string;
  usageCount: number;
  lastUsedAt?: string;
}

export interface SmartPhraseRequest {
  trigger: string;
  title: string;
  expansion: string;
  scope: SmartPhraseScope;
  hospitalId?: string;
  ownerUserId?: string;
  specialty?: string;
}

/**
 * Client wrapper around the /smart-phrases REST endpoints. Powers the
 * dot-phrase autocomplete in the per-section EncounterNote form (item 5)
 * and the SmartPhrase library admin screen.
 */
@Injectable({ providedIn: 'root' })
export class SmartPhraseService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = '/smart-phrases';

  autocomplete(prefix: string, hospitalId?: string | null): Observable<SmartPhrase[]> {
    let params = new HttpParams().set('prefix', prefix);
    if (hospitalId) {
      params = params.set('hospitalId', hospitalId);
    }
    return this.http.get<SmartPhrase[]>(`${this.baseUrl}/autocomplete`, { params });
  }

  getById(id: string): Observable<SmartPhrase> {
    return this.http.get<SmartPhrase>(`${this.baseUrl}/${id}`);
  }

  create(req: SmartPhraseRequest): Observable<SmartPhrase> {
    return this.http.post<SmartPhrase>(this.baseUrl, req);
  }

  update(id: string, req: SmartPhraseRequest): Observable<SmartPhrase> {
    return this.http.put<SmartPhrase>(`${this.baseUrl}/${id}`, req);
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  recordUsage(id: string): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/${id}/usage`, {});
  }
}
